sbi/simulators/simutils.py

```diff
@@ -134,6 +135,87 @@ def simulation_wrapper(
+                and outputs xs of shape (num_dim_x)
+            If simulation_batch_size > 1: takes in thetas of shape (simulation_batch_size, num_dim_theta)
+                and outputs xs of shape (simulation_batch_size, num_dim_x)
+        parameter_sample_fn: Function to call for generating theta, e.g. prior sampling
+        num_samples: Number of simulations to run
+        simulation_batch_size: Number of simulations that are run within a single batch
+            If simulation_batch_size == -1, we run a batch with all simulations required,
+            i.e. simulation_batch_size = num_samples
+        x_dim: dimensionality of a single simulator output
+
+    Returns: torch.Tensor parameters theta, torch.Tensor simulated data x
+    """
+
+    # generate theta by sampling from prior (round 1) or proposal (round > 1)
+    parameters = parameter_sample_fn(num_samples)
+
+    # if simulation_batch_size == -1, we run all simulations in a single batch
+    if simulation_batch_size == -1:
+        simulation_batch_size = num_samples
+
+    # initialize an empty array that stores the simulation outputs
+    # I opted to use numpy here since torch.Tensors are automatically converted to numpy,
+    # but np.arrays are not automatically converted to torch.Tensors
+    all_x_shape = np.concatenate(
+        (np.asarray([0], dtype=int), np.asarray(x_dim, dtype=int))
+    )
+    all_x = np.empty(all_x_shape, dtype=np.float32)
+
+    # count total number of simulations
+    n_accepted = 0
+
+    # loop over simulations # Todo: should we use jit here for speed-up?
+    while n_accepted < num_samples:
+
+        # make sure batch size doesn't exceed number of total simulations
+        n_batch = min(simulation_batch_size, num_samples - n_accepted,)
+
+        # run simulator
+        # if case is needed due to different behavior of simulation_batch_size == 1 and
+        # simulation_batch_size > 1. See docstring.
+        with torch.no_grad():
+            if simulation_batch_size > 1:
+                x = simulator(parameters[n_accepted : n_accepted + n_batch])
+            else:
+                # squeeze in case simulator provides an additional dimension for single parameters,
+                # e.g. in linearGaussian example. Then append a dimension to be able to concatenate.
+                x = np.squeeze(simulator(parameters[n_accepted]))[
+                    None,
+                ]
+
+        # add simulations to database of simulations. If simulator output x is torch.Tensor,
+        # it automatically gets converted to a np.array here.
+        all_x = np.concatenate((all_x, x), axis=0)
+
+        n_accepted += simulation_batch_size
+
+    # if simulation_batch_size >= num_samples, all_x might be a tensor already.
+    # In all other cases, it is a np.array, so we convert it to a tensor here.
+    if not isinstance(all_x, torch.Tensor):
+        all_x = torch.from_numpy(all_x)
+
+    return torch.tensor(parameters), all_x
+
+
```

Comment:
I think we dont need numpy here and we can shorten this function a bit. I happened to code up a function for abc smc yesterday that does something similar using only `pytorch`, maybe this is useful?": 
```
def _run_simulations(self, parameters: torch.Tensor, batch_size: int = 100):

        num_samples = parameters.shape[0]
        num_rest = num_samples % batch_size
        num_batches = (
            int(num_samples / (batch_size - num_rest))
            if num_samples >= batch_size
            else 0
        )

        data = []
        for batch_idx in range(num_batches):
            ps = parameters[
                batch_idx * batch_size : (batch_idx + 1) * batch_size,
            ]
            ds = self.simulation_fun(ps)
            data.append(ds)
            self.num_simulations += batch_size
            self.pbar.update(batch_size)

        # simulate the rest
        if num_rest > 0:
            ps = parameters[
                -num_rest:,
            ]
            ds = self.simulation_fun(ps)
            data.append(ds)
            self.num_simulations += num_rest
            self.pbar.update(num_rest)

        if self.num_simulations >= self.num_simulation_budget:
            raise SimulationBudgetExceeded()

        # return list of tensors collected into single tensor
        return torch.cat(data)
```

Reply:
I would also vote for changing the name of the function, as it is not really a simulator. it runs simulations, so, e.g., `simulate_in_batches`, or `_run_simulations_batch`

Reply:
Looks like `torch.chunk` could be used here? https://pytorch.org/docs/stable/torch.html#torch.chunk